File: kanz-py/kanz_bus/frame.py
"""Wire frame helper — :func:`unframe` for consumer-side decoding."""

from __future__ import annotations

from envelope.v1.envelope_pb2 import Envelope
from envelope.v1.event_frame_pb2 import EventFrame


def unframe(body: bytes) -> tuple[Envelope, bytes]:
    """Deserialize a ``bus.Message.body`` into ``(envelope, payload bytes)``.

    Consumers look up the payload schema in the registry (EVT-16) and decode
    the payload against it. The envelope is returned as-is; callers that
    want defensive validation should run :func:`kanz_bus.validate` themselves.

    Raises :class:`ValueError` on a malformed frame.
    """
    frame = EventFrame()
    try:
        frame.ParseFromString(body)
    except Exception as e:
        raise ValueError(f"frame unmarshal: {e}") from e
    if not frame.HasField("envelope"):
        raise ValueError("frame missing envelope")
    return frame.envelope, frame.payload

File: kanz/ts/kanz-bus/src/frame.ts
/** Wire frame helper — `unframe` for consumer-side decoding. */

import { fromBinary } from "@bufbuild/protobuf";
import type { Envelope } from "@kanz-eng/kanz-schemas/envelope/v1/envelope_pb.js";
import { EventFrameSchema } from "@kanz-eng/kanz-schemas/envelope/v1/event_frame_pb.js";

export interface Unframed {
  envelope: Envelope;
  payload: Uint8Array;
}

/**
 * Deserialize a `Message.body` into `(envelope, payload bytes)`.
 *
 * Consumers look up the payload schema in the registry (EVT-16) and
 * decode the payload against it. The envelope is returned as-is;
 * callers that want defensive validation should run {@link validate}
 * themselves.
 *
 * Throws on a malformed frame or a frame missing the envelope.
 */
export function unframe(body: Uint8Array): Unframed {
  let frame;
  try {
    frame = fromBinary(EventFrameSchema, body);
  } catch (e) {
    throw new Error(`frame unmarshal: ${(e as Error).message}`);
  }
  if (!frame.envelope) {
    throw new Error("frame missing envelope");
  }
  return { envelope: frame.envelope, payload: frame.payload };
}
